src/client/http/client.rs

```diff
@@ -92,6 +93,57 @@ impl HttpClient {
 		}
 	}
 
+	/// Perform a batch request towards the server.
+	///
+	/// Returns `Ok` if all requests were answered succesfully.
+	/// Returns `Error` if any of the requests fails.
+	//
+	// TODO(niklasad1): maybe simplify generic `requests`, it's quite unreadable.
+	pub async fn batch_request<'a>(
+		&self,
+		requests: impl IntoIterator<Item = (impl Into<String>, impl Into<jsonrpc::Params>)>,
+	) -> Result<Vec<JsonValue>, Error> {
+		let mut calls = Vec::new();
+		let mut ids = VecDeque::new();
+
+		for (method, params) in requests.into_iter() {
+			let id = self.request_id.fetch_add(1, Ordering::SeqCst);
+			calls.push(jsonrpc::Call::MethodCall(jsonrpc::MethodCall {
+				jsonrpc: jsonrpc::Version::V2,
+				method: method.into(),
+				params: params.into(),
+				id: jsonrpc::Id::Num(id),
+			}));
+			ids.push_back(id);
+		}
+
+		let batch_request = jsonrpc::Request::Batch(calls);
+		let response = self
+			.transport
+			.send_request_and_wait_for_response(batch_request)
+			.await
+			.map_err(|e| Error::TransportError(Box::new(e)))?;
+
+		match response {
+			jsonrpc::Response::Single(_) => {
+				Err(Error::Custom("Server replied with single response to a batch request".to_string()))
+			}
+			jsonrpc::Response::Notif(_notif) => {
+				Err(Error::Custom("Server replied with notification to a a batch request".to_string()))
+			}
+			jsonrpc::Response::Batch(rps) => {
+				let mut responses = Vec::with_capacity(ids.len());
+				for rp in rps {
+					let next_id = ids
+						.pop_front()
```

Comment:
This assumes the requests in the batch received in the same order as they are sent, maybe not a valid assumption note sure.

Reply:
Not sure what the server-side code for this looks like, but the jsonrpc spec (from [jsonrpc.org](https://www.jsonrpc.org/specification#batch)) says:
> The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object.

So in the end I guess it's an implementation detail depending on how jsonrpsee library processes the requests

Everything else LGTM though

Random question though because i'm not totally sure on the specifics, do batch requests work for HTTP/IPC/WS or is it just over http?

Reply:
Ok, cool missed that completely. I will fix that tomorrow.

Reply:
After implementing requests in any order I just realized that the user can't know whether we receive the requests in order or not, thus we need to sort the responses by ID before returning them to the user.

Reply:
Pretty sure that is not a valid assumption. 
Maybe you need to use a `BTreeMap` here between ids and requests/responses?

Reply:
Indeed, it's already implemented with a Vec instead of BTreeMap.

The reason is that the request IDs may wrap around and we can't order them by ID then